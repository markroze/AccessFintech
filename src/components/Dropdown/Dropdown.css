.dropdown {
  position: relative;
  display: inline-block;
  text-align: left;
}
.dropdown__button {
  border: none;
}
.dropdown__popover {
  padding: 0;
  margin: 0;
  position: absolute;
}
.dropdown__popover__item {
  min-width: 90px;
  padding: 0.5rem;
  cursor: pointer;
  white-space: nowrap;
}
.dropdown__popover__item:hover {
  background-color: hsla(210, 100%, 33%, 0.5);
}
