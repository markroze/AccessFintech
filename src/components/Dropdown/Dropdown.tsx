import React, { useCallback, useEffect, useState } from 'react';
import './Dropdown.css';
import MoreIcon from './../MoreIcon';
import Theme from '../../theme';

const dropdownBaseStyles = () => ({
  backgroundColor: Theme.colors.black600,
  color: Theme.colors.white100,
});

export interface DropdownOption {
  value: string;
  label: string;
}

interface DropdownProps {
  options: DropdownOption[];
  onOptionSelect?: (option: DropdownOption) => void;
  className?: string;
  tabIndex?: number;
  size?: 'samll' | 'medium' | 'large';
  position?: 'left' | 'right';
}

const Dropdown = ({
  options,
  onOptionSelect,
  className,
  tabIndex = 0,
  size = 'medium',
  position = 'right',
}: DropdownProps) => {
  const [selectedItem, setSelectedItem] = useState<DropdownOption | null>(null);
  const [isOpen, setIsOpen] = useState(false);
  const [activeIndex, setActiveIndex] = useState(0);

  const handleClose = () => {
    setIsOpen(false);
    setActiveIndex(0);
  };

  const handleOptionSelect = useCallback(
    (option: DropdownOption) => {
      setSelectedItem(option);
      handleClose();
      onOptionSelect && onOptionSelect(option);
    },
    [onOptionSelect]
  );

  const listener = useCallback(
    (event: KeyboardEvent) => {
      if (event.code === 'ArrowDown') {
        setActiveIndex((prev) => (prev + 1) % options.length);
      }
      if (event.code === 'ArrowUp') {
        setActiveIndex((prev) => (prev - 1 + options.length) % options.length);
      }

      if (event.code === 'Escape') {
        handleClose();
      }
      if (event.code === 'Space' || event.code === 'Enter') {
        handleOptionSelect(options[activeIndex]);
      }
      event.preventDefault();
    },
    [options, activeIndex, handleOptionSelect]
  );

  useEffect(() => {
    if (isOpen) {
      document.addEventListener('keydown', listener);
    }
    return () => {
      document.removeEventListener('keydown', listener);
    };
  }, [listener, isOpen]);

  return (
    <div
      {...{
        className: `dropdown ${className}`,
      }}
    >
      <button
        {...{
          tabIndex,
          className: 'dropdown__button',
          style: dropdownBaseStyles(),
          'aria-haspopup': true,
          'aria-expanded': isOpen,
          'aria-labelledby': 'dropdown__button',
          onClick: () => setIsOpen(!isOpen),
        }}
      >
        <MoreIcon />
      </button>
      {isOpen && (
        <ul
          {...{
            className: `dropdown__popover`,
            style: {
              ...dropdownBaseStyles(),
              ...(position === 'left' ? { left: 0 } : { right: 0 }),
            },
          }}
        >
          {options.map((option, ind) => (
            <li
              {...{
                tabIndex: tabIndex + ind,
                key: ind,
                value: option.value,
                className: 'dropdown__popover__item',
                onClick: () => handleOptionSelect(option),
                'aria-selected': selectedItem?.value === option.value,
                role: 'option',
                style: {
                  ...(selectedItem?.value === option.value && {
                    backgroundColor: Theme.colors.blue300,
                  }),
                  ...(activeIndex === ind && {
                    backgroundColor: 'hsla(210, 100%, 33%, 0.5)',
                  }),
                },
              }}
            >
              {option.label}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Dropdown;
