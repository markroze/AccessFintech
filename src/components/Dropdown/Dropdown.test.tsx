import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Dropdown from './Dropdown';

const users = [
  { value: 'mark', label: 'Mark' },
  { value: 'jane', label: 'Jane' },
  { value: 'john', label: 'John' },
];

describe('Dropdown', () => {
  it('should render correctly', () => {
    const { container } = render(<Dropdown options={users} />);
    expect(container).toMatchSnapshot();
  });

  it('should open when clicked on button', async () => {
    render(<Dropdown options={users} />);
    const dropdownButton = screen.getByRole('button');
    await userEvent.click(dropdownButton);
    const dropdownMenu = screen.getByTestId('dropdownItems');
    expect(dropdownMenu).toBeInTheDocument();
  });

  it('should close when clicked on button', async () => {
    render(<Dropdown options={users} />);
    const dropdownButton = screen.getByRole('button');

    await userEvent.click(dropdownButton);
    const dropdownMenu = screen.getByTestId('dropdownItems');
    expect(dropdownMenu).toBeInTheDocument();
    await userEvent.click(dropdownButton);
    expect(dropdownMenu).not.toBeInTheDocument();
  });
});
