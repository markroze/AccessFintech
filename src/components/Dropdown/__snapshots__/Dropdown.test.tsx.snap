// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<div>
  <div
    class="dropdown undefined"
    datatestid="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-labelledby="dropdown__button"
      class="dropdown__button"
      style="background-color: rgb(48, 48, 48); color: rgb(255, 255, 255);"
      tabindex="0"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M10.5 16.5a1.5 1.5 0 103 0 1.5 1.5 0 00-3 0zm0-4.5a1.5 1.5 0 103 0 1.5 1.5 0 00-3 0zm0-4.5a1.5 1.5 0 103 0 1.5 1.5 0 00-3 0z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
